### PR TITLE
Add design doc for waiters

### PR DESCRIPTION
**Pull request by zoewangg**
## Description
Add design doc for waiters #24 

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: docs/design/core/waiters/README.md [hunk 1]
```diff
@@ -0,0 +1,447 @@
...
+
+    /**
+     * Poller method that waits for the table status to transition to <code>ACTIVE</code> by
+     * invoking {@link DynamoDbClient#describeTable}. It returns when the resource enters into a desired state or
+     * it is determined that the resource will never enter into the desired state.
+     *
+     * @param describeTableRequest Represents the input of a <code>DescribeTable</code> operation.
+     * @return {@link DescribeTableResponse}
+     */
+    default DescribeTableResponse waitUtilTableExists(DescribeTableRequest describeTableRequest) {
```
Comment (millems) [inline docs/design/core/waiters/README.md:87]: Until*?

Comment (zoewangg) [reply docs/design/core/waiters/README.md:87]: Oops! I blame intellij for not telling me.

File: docs/design/core/waiters/README.md [hunk 2]
```diff
@@ -0,0 +1,447 @@
...
+```
+
+### `WaiterAcceptor`
+
+`WaiterAcceptor` is a class that inspects the response or error returned from the operation and determines whether an expected condition
+is met and indicates the next state that the waiter should be transitioned to if there is a match.
+
+```java
+@SdkPublicApi
+public interface WaiterAcceptor<T> {
```
Comment (millems) [inline docs/design/core/waiters/README.md:302]: Can we bikeshed on the name?

File: docs/design/core/waiters/README.md [hunk 3]
```diff
@@ -0,0 +1,447 @@
...
+            public WaiterState waiterState() {
+                return WaiterState.RETRY;
+            }
+
+            @Override
+            public boolean matches(Throwable t) {
+                return errorPredicate.test(t);
+            }
+        };
+    }
```
Comment (millems) [inline docs/design/core/waiters/README.md:390]: Isn't retrying on an exception just as common as retrying on a non-exception?

File: docs/design/core/waiters/README.md [hunk 4]
```diff
@@ -0,0 +1,447 @@
...
+1. slightly better discoverability
+
+**Decision:** Option 1 will be used, because it is consistent with existing features and option2 might bloat the size
+of the client, making it more difficult to use.
+
+### Why returning the service response for waiter operations with a success state of a specific response?
+
+The reason that the last response that has satisfied the waiter success state is returned is because it is a common pattern that customers creates a resource and then
+retrieves the information of the resources. Without returning the response, customers will have to send an extra request when the waiter returns. 
+See [feature request](https://github.com/aws/aws-sdk-java/issues/815)
```
Comment (millems) [inline docs/design/core/waiters/README.md:434]: Can we cache *every* response, so they can see what happened, instead of just the latest?

**Pull request merged**